File: 36.cpp
struct Solution {
    bool isValidSudoku(vector<vector<char>>& board) {
        short r[9] = {0}, c[9] = {0}, b[9] = {0};
        for (int i = 0; i < 9; ++i){
            for (int j = 0; j < 9; ++j){
                char x = board[i][j];
                if (x == '.'){
                    continue;
                }
                int m = 1 << (x - '1'), k = (i / 3) * 3 + (j / 3);
                if (((r[i] | c[j] | b[k]) & m) != 0){
                    return false;
                }
                r[i] |= m;
                c[j] |= m;
                b[k] |= m;
            }
        }
        return true;
    }
};

File: sol/36/36.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        r = [0] * 9
        c = [0] * 9
        b = [0] * 9
        md = {str(i): 1 << i for i in range(10)}
        q = tuple(i // 3 for i in range(9))
        for i in range(9):
            for j in range(9):
                x = board[i][j]
                if x == '.':
                    continue
                m = md[x]
                d = q[i] * 3 + q[j]
                if ((r[i] | c[j] | b[d]) & m) != 0:
                    return False
                r[i] |= m
                c[j] |= m
                b[d] |= m
        return True
